src/libopensc/card-skeid4.c

```diff
@@ -0,0 +1,461 @@
+
+	if (card == NULL || env == NULL || env->key_ref_len != 1) {
+		return SC_ERROR_INVALID_ARGUMENTS;
+	}
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_3_SHORT, 0x22, 0x41, 0);
+
+	if (env->operation == SC_SEC_OPERATION_DECIPHER) {
+		apdu.p2 = 0xB8;
+		algo_ref = 0x1A;
+	}
+	else if (env->operation != SC_SEC_OPERATION_SIGN)
+		return SC_ERROR_INVALID_ARGUMENTS;
+	else if (*env->key_ref == 0x60) {
+		apdu.p2 = 0xA4;
+		algo_ref = 2;
+	}
+	else {
+		apdu.p2 = 0xB6;
+		algo_ref = 0x42;
+	}
+
+	p = sbuf;
+
+	*p++ = 0x84; /* key reference */
+	*p++ = 1;
+	*p++ = *env->key_ref;
+
+	*p++ = 0x80;	/* algorithm reference */
+	*p++ = 0x01;
+	*p++ = algo_ref;
+
+	apdu.lc = 6;
+	apdu.datalen = 6;
+	apdu.data = sbuf;
+
+	r = sc_transmit_apdu(card, &apdu);
+	if (r) {
+		sc_log(card->ctx, "%s: APDU transmit failed", sc_strerror(r));
+		goto err;
+	}
+	r = sc_check_sw(card, apdu.sw1, apdu.sw2);
+	if (r) {
+		sc_log(card->ctx, "%s: Card returned error", sc_strerror(r));
+		goto err;
+	}
+
+	priv->sec_env = env ;
+	r = SC_SUCCESS;
+err:
+	return r;
+}
+
+static int skeid4_select_file(struct sc_card *card, const struct sc_path *in_path, struct sc_file **file_out)
+{
+	// adapted from iso7816_select_file
+
+	struct sc_context *ctx;
+	struct sc_apdu apdu;
+	unsigned char buf[SC_MAX_APDU_BUFFER_SIZE];
+	unsigned char pathbuf[SC_MAX_PATH_SIZE], *path = pathbuf;
+	int r, pathlen, pathtype;
+	int select_mf = 0;
+	struct sc_file *file = NULL;
+	const u8 *buffer;
+	size_t buffer_len;
+	unsigned int cla, tag;
+
+	if (card == NULL || in_path == NULL) {
+		return SC_ERROR_INVALID_ARGUMENTS;
+	}
+	ctx = card->ctx;
+	memcpy(path, in_path->value, in_path->len);
+	pathlen = in_path->len;
+	pathtype = in_path->type;
+
+	if (in_path->aid.len) {
+		if (!pathlen) {
+			memcpy(path, in_path->aid.value, in_path->aid.len);
+			pathlen = in_path->aid.len;
+			pathtype = SC_PATH_TYPE_DF_NAME;
+		} else {
+			/* First, select the application */
+			sc_format_apdu(card, &apdu, SC_APDU_CASE_3_SHORT, 0xA4, 4, 0);
+			apdu.data = in_path->aid.value;
+			apdu.datalen = in_path->aid.len;
+			apdu.lc = in_path->aid.len;
+
+			r = sc_transmit_apdu(card, &apdu);
+			LOG_TEST_RET(ctx, r, "APDU transmit failed");
+			r = sc_check_sw(card, apdu.sw1, apdu.sw2);
+			if (r)
+				LOG_FUNC_RETURN(ctx, r);
+
+			if (pathtype == SC_PATH_TYPE_PATH
+					|| pathtype == SC_PATH_TYPE_DF_NAME)
+				pathtype = SC_PATH_TYPE_FROM_CURRENT;
+		}
+	}
+
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_4_SHORT, 0xA4, 0, 0);
+
+	switch (pathtype) {
+	case SC_PATH_TYPE_FILE_ID:
+		apdu.p1 = 2;
+		if (pathlen != 2)
+			return SC_ERROR_INVALID_ARGUMENTS;
+		break;
+	case SC_PATH_TYPE_DF_NAME:
+		apdu.p1 = 4;
+		break;
+	case SC_PATH_TYPE_PATH:
+		apdu.p1 = 8;
+		if (pathlen >= 2 && memcmp(path, "\x3F\x00", 2) == 0) {
+		        apdu.p1 = 2;
+			if (pathlen == 2) {	/* only 3F00 supplied */
+				select_mf = 1;
+				break;
+			}
+			path += 2;
+			pathlen -= 2;
+		}
+		break;
+	case SC_PATH_TYPE_FROM_CURRENT:
+		apdu.p1 = 2;
```

Comment:
In ISO/IEC 7816, 0x02 only references a file identifier under the current DF, so this may not work with a longer path, which is why the iso7816 uses 0x09